### PR TITLE
[FEATURE] Lire les données des CGU avec ou sans la nouvelle api (PIX-22580)

### PR DESCRIPTION
**Pull request by EmmanuelleBonnemay**
## 💥 Problème

Dans le cadre de la nouvelle gestion du versioning des CGUs pour Pix App, nous avons besoin d’une API permettant de récupérer l'état d’acceptation d’un document légal pour un utilisateur.

## 👩‍🚀 Proposition

À faire dans le usecase `getLegalDocumentByUserId` (ajouter la gestion du feature toggle `on` ou `off`) 

CAS 1 : Si le feature toggle `newPixAppLegalDocumentsVersioning` est désactivé et `service=pix-app` et `type=TOS`.

On doit utiliser l’ancien système des CGU (table `users`) pour faire la correspondance avec le nouveau système (`legal document`), selon les règles suivantes :
    Si `cgu=true` et `mustValidateTermsOfService=true`, alors `status=update-requested`
    Si `cgu=true` et `mustValidateTermsOfService=false`, alors `status=accepted`
    Si `cgu=false` alors `status=requested` SAUF POUR LES ELEVES SCO alors `status=skipped`. 

 CAS 2 : Si le feature toggle `newPixAppLegalDocumentsVersioning` est activé.
    Vérifier la table d’acceptation pour l’utilisateur (déjà présent aujourd’hui)
    Si (status=requested ou `update-requested` et `service=pix-app` et `type=TOS`), renvoyer le `status=skipped` si c’est un élève SCO.

Il faudra gérer la nouvelle valeur `skipped` et rajouter une méthode `buildForLegacyPixAppCgu` dans le modèle `legalDocumentStatus`

 **Règles pour définir un élève SCO** 
            tout user ayant au moins un `orga-learner` avec une date de création <= 6 ans
            dans une organisation de type SCO et qui a `isManagingStudents` à `true`
            sans demande de sortie du SCO (table `account-recovery-demands` à `true`)
    `acceptedAt` equivalent à `lastTermsOfServiceValidatedAt`
 
## 👁️ Remarques

- Cas d'un élève qui serait learner de plusieurs organisations et qui aurait fait une demande de récupération de compte (associée à l'un des learners ?) > comment doit se passer le calcul de isStudent dans ce cas ?
- Un ticket "Modifier l'appel Pix App qui récupère le statut des CGU afin de fonctionner avec le nouveau modèle" est prévu prochainement, cette PR sera testable à ce moment là.

## ♻️ Pour tester

Cette PR ne doit pas avoir d'impact sur notre code de production.

Tester la non régression sur Pix App avec le FT activé et non activé : 
* Pour demander la validation pour de nouvelles CGU avec le FT activé : 
   ```shell
   node src/legal-documents/scripts/add-new-legal-document-version.js --type=TOS --service=pix-app --versionAt=2026-05-05
   ```
* Pour demander la validation pour de nouvelles CGU avec le FT non-activé : 
   ```sql
   update "users" set "mustValidateTermsOfService"=true where email='justin.compte@example.net';
   ```

### CODE REVIEW

File: api/src/legal-documents/domain/usecases/get-legal-document-status-by-user-id.usecase.js [hunk 1]
```diff
@@ -12,17 +13,44 @@ import { LegalDocumentType } from '../models/LegalDocumentType.js';
...
+  if (!isNewVersioningEnabled && isPixAppTos) {
+    const user = await userRepository.findPixAppLegacyCguByUserId(userId);
+    const legalDocumentStatus = LegalDocumentStatus.buildForLegacyPixAppCgu({
+      cgu: user.cgu,
+      mustValidateTermsOfService: user.mustValidateTermsOfService,
+      lastTermsOfServiceValidatedAt: user.lastTermsOfServiceValidatedAt,
+    });
+
+    if (legalDocumentStatus.status === STATUS.REQUESTED) {
+      const isScoStudent = await organizationLearnerRepository.isUserScoStudent({ userId });
```
Comment (EmmanuelleBonnemay) [inline api/src/legal-documents/domain/usecases/get-legal-document-status-by-user-id.usecase.js]: Si le feature toggle n'est pas activé nous calculons le `isScoStudent` seulement si le statut est `REQUESTED`.
Si le feature toggle est activé nous calculons le `isScoStudent` si le statut est `REQUESTED` ou `UPDATE_REQUESTED.`
Cette différence s'explique parce que le premier cas s'appuie sur le `buildForLegacyPixAppCgu` qui pose le `UPDATE_REQUESTED` en s'appuyant sur `mustValidateTermsOfService` = true (dans la table users). 
Or ce `mustValidateTermsOfService` = true n'est posé que sur les utilisateurs ayant `cgu` à `true` ce qui n'est pas le cas des élèves (cette opération est faite par le script  `mark-users-for-TermsOfService-revalidation`).
Conclusion : dans l'ancienne implémentation (FT = false), aucun élève n'aura le statut `UPDATE_REQUESTED`.

File: api/tests/legal-documents/integration/domain/usecases/get-legal-document-status-by-user-id.usecase.test.js [hunk 2]
```diff
@@ -54,4 +59,219 @@ describe('Integration | Legal documents | Domain | Use case | get-legal-document
...
+
+        // when
+        const result = await usecases.getLegalDocumentStatusByUserId({ userId: user.id, service: PIX_APP, type: TOS });
+
+        // then
+        expect(result).to.be.an.instanceOf(LegalDocumentStatus);
+        expect(result.status).to.equal(STATUS.SKIPPED);
+      });
+
+      it('returns accepted status when user has accepted the current version even if SCO student', async function () {
```
Comment (EmmanuelleBonnemay) [inline api/tests/legal-documents/integration/domain/usecases/get-legal-document-status-by-user-id.usecase.test.js]: cette situation ne "devrait" jamais se produire, voir si le test est nécessaire et le cas échéant, si nous voulons ce comportement.

File: api/tests/legal-documents/integration/domain/usecases/get-legal-document-status-by-user-id.usecase.test.js [hunk 3]
```diff
@@ -54,4 +59,219 @@ describe('Integration | Legal documents | Domain | Use case | get-legal-document
...
+
+        // when
+        const result = await usecases.getLegalDocumentStatusByUserId({ userId: user.id, service: PIX_APP, type: TOS });
+
+        // then
+        expect(result).to.be.an.instanceOf(LegalDocumentStatus);
+        expect(result.status).to.equal(STATUS.SKIPPED);
+      });
+
+      it('returns skipped status when user has an older acceptance and is an SCO student', async function () {
```
Comment (EmmanuelleBonnemay) [inline api/tests/legal-documents/integration/domain/usecases/get-legal-document-status-by-user-id.usecase.test.js]: Cette situation ne "devrait" pas se produire, voir si le test est nécessaire et le cas échéant, si nous voulons ce comportement.

File: api/src/prescription/organization-learner/infrastructure/repositories/organization-learner-repository.js [hunk 4]
```diff
@@ -504,6 +504,30 @@ const findAllLearnerWithAtLeastOneParticipationByOrganizationIds = async functio
...
+    .where({ 'organizations.isManagingStudents': true, 'organizations.type': 'SCO' })
+    .where({ 'organization-learners.userId': userId, 'organization-learners.deletedAt': null })
+    .where('organization-learners.createdAt', '>=', sixYearsAgo)
+    .first();
+
+  if (!scoLearner) return false;
+
+  const accountRecoveryDemand = await knexConnection('account-recovery-demands')
+    .where({ userId, organizationLearnerId: scoLearner.learnerId, used: true })
+    .first();
```
Comment (lego-technix) [inline api/src/prescription/organization-learner/infrastructure/repositories/organization-learner-repository.js]: Placer cette fonction `isUserScoStudent` dans l’_organizationLearnerRepository_ ne semble pas satisfaisant. En effet : 
* un repository doit être le plus « dummy » possible
* cette fonction `isUserScoStudent` fait une requête sur la table `account-recovery-demands` qui n’a rien à voir avec l’_organizationLearnerRepository_

Aussi je suggère de mettre cette fonction `isUserScoStudent` plutôt dans un modèle ou un service. À discuter ensemble.

File: api/src/prescription/organization-learner/infrastructure/repositories/organization-learner-repository.js
Comment (lego-technix) [inline api/src/prescription/organization-learner/infrastructure/repositories/organization-learner-repository.js:1]: L’ajout de la fonctionnalité `isUserScoStudent` mérite son commit dédié et ne me semble pas appartenir à un commit ayant pour sujet `add legacy cgu handling in repositories`, en effet il me semble que la fonctionnalité `isUserScoStudent` va être utilisée dans d’autres contextes qu’uniquement la gestion des CGU legacy.

File: api/src/legal-documents/infrastructure/repositories/user-sco.repository.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import { DomainTransaction } from '../../../shared/domain/DomainTransaction.js';
+
+const SIX_YEARS_IN_MS = 6 * 365.25 * 24 * 60 * 60 * 1000;
+
+export const isUserScoStudent = async ({ userId }) => {
```
Comment (lego-technix) [inline api/src/legal-documents/infrastructure/repositories/user-sco.repository.js]: Suggestion : comme il y a un unique paramètre, il est mieux de ne pas utiliser une structure d’objet pour le passer :
```suggestion
export const isUserScoStudent = async (userId) => {
```

Comment (lego-technix) [inline api/src/legal-documents/infrastructure/repositories/user-sco.repository.js]: Suggestion : utiliser un marqueur dédié pour décourager l’utilisation de cette  fonction (le marqueur `@deprecated` est très efficacement montré comme étant à éviter par tous les IDE) : 
```suggestion
/**
  * @deprecated avoid using this function as much as possible because it has performance issues and because it should be obtained from a Prescription internal API when it is available.
  */
export const isUserScoStudent = async ({ userId }) => {
```

File: api/src/prescription/organization-learner/infrastructure/repositories/organization-learner-repository.js (hunk 4 above)
Comment (lego-technix) [reply api/src/prescription/organization-learner/infrastructure/repositories/organization-learner-repository.js]: Le nouvel emplacement utilisé est satisfaisant.

File: api/src/prescription/organization-learner/infrastructure/repositories/organization-learner-repository.js
Comment (lego-technix) [reply api/src/prescription/organization-learner/infrastructure/repositories/organization-learner-repository.js:1]: Il y a maintenant un commit dédié. C’est parfait.

File: api/tests/legal-documents/integration/infrastructure/repositories/user-sco.repository.test.js [hunk 6]
```diff
@@ -0,0 +1,183 @@
+import { isUserScoStudent } from '../../../../../src/legal-documents/infrastructure/repositories/user-sco.repository.js';
+import { expect } from '../../../../test-helper.js';
+import { databaseBuilder } from '../../../../tooling/databases.js';
+
+describe('Integration | Infrastructure | Repository | Organization Learner', function () {
```
Comment (lego-technix) [inline api/tests/legal-documents/integration/infrastructure/repositories/user-sco.repository.test.js]:
```suggestion
describe('Integration | Infrastructure | Repository | UserScoRepository', function () {
```

File: api/src/legal-documents/infrastructure/repositories/user.repository.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+import { UserNotFoundError } from '../../../shared/domain/errors.js';
+
+const findPixAppLegacyCguByUserId = async (userId) => {
+  const knexConnection = DomainTransaction.getConnection();
+  const user = await knexConnection('users')
+    .select('cgu', 'mustValidateTermsOfService', 'lastTermsOfServiceValidatedAt')
+    .where({ id: userId })
+    .first();
+  if (!user) throw new UserNotFoundError();
+  return user;
```
Comment (lego-technix) [inline api/src/legal-documents/infrastructure/repositories/user.repository.js]: La fonction s’appelle ici `findPixAppLegacyCguByUserId`, or d’après les standards un `find` implique qu’on ne renvoie pas d’erreur si aucun résultat n’est trouvé. Si on veut que l’absence de résultat trouvé renvoie une erreur, alors c’est un `get` qu’il faut utiliser : `getPixAppLegacyCguByUserId`

File: api/tests/legal-documents/integration/infrastructure/repositories/user.repository.test.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
+import * as userRepository from '../../../../../src/legal-documents/infrastructure/repositories/user.repository.js';
+import { UserNotFoundError } from '../../../../../src/shared/domain/errors.js';
+import { expect } from '../../../../test-helper.js';
+import { databaseBuilder } from '../../../../tooling/databases.js';
+import { catchErr } from '../../../../tooling/test-utils/error.js';
+
+describe('Integration | Legal documents | Infrastructure | Repository | PixAppCgu', function () {
```
Comment (lego-technix) [inline api/tests/legal-documents/integration/infrastructure/repositories/user.repository.test.js]:
```suggestion
describe('Integration | Legal documents | Infrastructure | Repository | UserRepository', function () {
```

File: api/src/legal-documents/domain/usecases/get-legal-document-status-by-user-id.usecase.js (hunk 1 above)
Comment (EmmanuelleBonnemay) [reply api/src/legal-documents/domain/usecases/get-legal-document-status-by-user-id.usecase.js]: Ce commentaire n'est plus pertinent après les dernières modifications

File: api/tests/legal-documents/integration/domain/usecases/get-legal-document-status-by-user-id.usecase.test.js (hunk 3 above)
Comment (EmmanuelleBonnemay) [reply api/tests/legal-documents/integration/domain/usecases/get-legal-document-status-by-user-id.usecase.test.js]: Cette situation est possible dans le cas d'un élève ayant accepté d'anciennes cgu avant de devenir "sco" : dans ce cas son statut est "update-requested" , et le usecase vérifie s'il est sco ou non ; comme il est sco, son statut est modifié et devient `NOT_APPLICABLE`.
Conclusion : le test est nécessaire et couvre un edge-case.

Comment (EmmanuelleBonnemay) [reply api/tests/legal-documents/integration/domain/usecases/get-legal-document-status-by-user-id.usecase.test.js]: C'est bien le comportement que nous voulons.

File: api/tests/legal-documents/integration/domain/usecases/get-legal-document-status-by-user-id.usecase.test.js (hunk 2 above)
Comment (EmmanuelleBonnemay) [reply api/tests/legal-documents/integration/domain/usecases/get-legal-document-status-by-user-id.usecase.test.js]: Cette situation est possible dans le cas d'un élève ayant accepté les cgu courantes avant de devenir "sco" : dans ce cas on ne lui propose pas de compléter les cgu car son statut est "accepted" (on ne vérifie pas son statut de sco). Son statut sco sera vérifié lors de la connexion qui suivra la mise à jour des cgu.
Conclusion : le test est nécessaire et couvre un edge-case.

File: api/src/legal-documents/infrastructure/repositories/user-sco.repository.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
+import { DomainTransaction } from '../../../shared/domain/DomainTransaction.js';
+
+const SIX_YEARS_IN_MS = 6 * 365.25 * 24 * 60 * 60 * 1000;
+
+/**
+ * @deprecated avoid using this function as much as possible because it has performance issues and because it should be obtained from a Prescription internal API when it is available.
+ */
+
+export const isUserScoStudent = async (userId) => {
```
Comment (lego-technix) [inline api/src/legal-documents/infrastructure/repositories/user-sco.repository.js:8]: Cosmétique : Coller la JSDoc au code documenté : 
```suggestion
/**
 * @deprecated avoid using this function as much as possible because it has performance issues and because it should be obtained from a Prescription internal API when it is available.
 */
export const isUserScoStudent = async (userId) => {
```

File: api/tests/legal-documents/integration/domain/usecases/get-legal-document-status-by-user-id.usecase.test.js [hunk 10]
```diff
@@ -54,4 +59,187 @@ describe('Integration | Legal documents | Domain | Use case | get-legal-document
...
     });
   });
+
+  context('when service is pix-app', function () {
+    context('when feature toggle is disabled', function () {
+      beforeEach(async function () {
+        await featureToggles.set('newPixAppLegalDocumentsVersioning', false);
+      });
+
+      it('returns accepted status when and mustValidateTermsOfService=false', async function () {
```
Comment (lego-technix) [inline api/tests/legal-documents/integration/domain/usecases/get-legal-document-status-by-user-id.usecase.test.js]: Pas bloquant : dans ce fichier de test les « when » sont à la fin des « it » : _« when and mustValidateTermsOfService=false »_, _« when mustValidateTermsOfService=true »_, etc. Il serait mieux d’utiliser des `context` pour cela.

**Review by lego-technix — COMMENTED** [review]
Revue du code effectuée et positive, sauf en ce qui concerne la grosse requête SQL de `isUserScoStudent` qui va me demander plus de temps.

File: api/src/legal-documents/domain/usecases/get-legal-document-status-by-user-id.usecase.js [hunk 11]
```diff
@@ -12,27 +13,60 @@ import { LegalDocumentType } from '../models/LegalDocumentType.js';
...
+    (legalDocumentStatus.status === STATUS.REQUESTED || legalDocumentStatus.status === STATUS.UPDATE_REQUESTED)
+  ) {
+    const isScoStudent = await userScoRepository.isUserScoStudent(userId);
+    if (isScoStudent) {
+      return new LegalDocumentStatus({
+        status: STATUS.NOT_APPLICABLE,
+        acceptedAt: null,
+        documentPath: legalDocumentStatus.documentPath,
+      });
+    }
```
Comment (bpetetot) [inline api/src/legal-documents/domain/usecases/get-legal-document-status-by-user-id.usecase.js:66]: Si c'est non applicable, on a pas besoin de retourner de `documentPath`
```
documentPath: null,
```

**Review by bpetetot — APPROVED** [review]
✅ Excellent ! Juste une petite remarque à prendre en compte.

File: api/src/legal-documents/infrastructure/repositories/user-sco.repository.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+      'organization-learners.userId': userId,
+      'organization-learners.deletedAt': null,
+    })
+    .where('organization-learners.createdAt', '>=', sixYearsAgo)
+    .whereNotExists(function () {
+      this.select('*')
+        .from('account-recovery-demands')
+        .where({ userId, used: true })
+        .whereRaw('"organizationLearnerId" = "organization-learners"."id"');
+    })
```
Comment (lego-technix) [inline api/src/legal-documents/infrastructure/repositories/user-sco.repository.js:27]: Petite optimisation : 
```suggestion
    .whereNotExists(function () {
      this.select(1)
        .from('account-recovery-demands')
        .where({ userId, used: true })
        .whereRaw('"organizationLearnerId" = "organization-learners"."id"');
    })
```

**Review by lego-technix — APPROVED** [review]
✅ Lu et testé fonctionnellement avec succès avec Firefox 🦊 et Linux 🐧
